### ISSUE TITLE
Windows Mingw build has warnings about threading not supported

### ISSUE THREAD

**Issue by rgommers** [rgommers/scipy#125]
These warnings seem to indicate a real issue:

```
[94/1557] Compiling C object scipy/linalg/_flapack.cp39-win_amd64.pyd.p/meson-generated_..__flapackmodule.c.obj
In file included from C:\hostedtoolcache\windows\Python\3.9.9\x64\lib\site-packages\numpy\core\include/numpy/ndarraytypes.h:4,
                 from C:\hostedtoolcache\windows\Python\3.9.9\x64\lib\site-packages\numpy\core\include/numpy/ndarrayobject.h:12,
                 from C:\hostedtoolcache\windows\Python\3.9.9\x64\lib\site-packages\numpy\core\include/numpy/arrayobject.h:4,
                 from C:\hostedtoolcache\windows\Python\3.9.9\x64\lib\site-packages\numpy\f2py\src/fortranobject.h:13,
                 from scipy/linalg/_flapackmodule.c:16:
scipy/linalg/_flapackmodule.c:1248:1: warning: 'thread' attribute directive ignored [-Wattributes]
 1248 | static F2PY_THREAD_LOCAL_DECL cb_cselect_in_gees__user__routines_t *_active_cb_cselect_in_gees__user__routines = NULL;
      | ^~~~~~
scipy/linalg/_flapackmodule.c:1410:1: warning: 'thread' attribute directive ignored [-Wattributes]
 1410 | static F2PY_THREAD_LOCAL_DECL cb_zselect_in_gees__user__routines_t *_active_cb_zselect_in_gees__user__routines = NULL;
      | ^~~~~~
scipy/linalg/_flapackmodule.c:1572:1: warning: 'thread' attribute directive ignored [-Wattributes]
 1572 | static F2PY_THREAD_LOCAL_DECL cb_sselect_in_gees__user__routines_t *_active_cb_sselect_in_gees__user__routines = NULL;
      | ^~~~~~
scipy/linalg/_flapackmodule.c:1738:1: warning: 'thread' attribute directive ignored [-Wattributes]
 1738 | static F2PY_THREAD_LOCAL_DECL cb_dselect_in_gees__user__routines_t *_active_cb_dselect_in_gees__user__routines = NULL;
      | ^~~~~~

...
```

There are 30 such warnings, in `linalg/_interpolativemodule.c`, `linalg/_flapackmodule.c`, `linalg/_propack/_[c|s|d|z]propackmodule.c`, `integrate/_lsodamodule.c`, `integrate/_vodemodule.c`, `integrate/_dopmodule.c`, and `optimize/_cobylamodule.c`.

Probably relevant links:
- https://bugs.freedesktop.org/show_bug.cgi?id=57591
- https://sourceforge.net/p/mingw-w64/bugs/327/
- https://github.com/numpy/numpy/issues/19437

From the second link: _"Because `__declspec(thread)` results in just a warning and not a compilation failure, the configure script thinks that __declspec(thread) is supported and happens to select it"_. If that's the case here too, we may have a problem.

The code generated by `f2py` looks like:
```
#ifndef F2PY_THREAD_LOCAL_DECL
#if defined(_MSC_VER) \
      || defined(_WIN32) || defined(_WIN64) \
      || defined(__MINGW32__) || defined(__MINGW64__)
#define F2PY_THREAD_LOCAL_DECL __declspec(thread)
```
So it explicitly deals with mingw-w64, but it looks like mingw-w64 doesn't recognize what `f2py` is doing here.

Cc @matthew-brett, @carlkl

Also Cc @pearu in case this rings a bell.

**Comment by pearu** [comment]
> So it explicitly deals with mingw-w64, but it looks like mingw-w64 doesn't recognize what f2py is doing here.

This should be the other way around: f2py should do what mingw-w64 dictates it to do.

So, the question is, what is the correct threading directive in the case of mingw-w64? https://sourceforge.net/p/mingw-w64/bugs/327/ seems to suggest `__thread` and if this is correct, the fix would be to move `defined(__MINGW64__)` to the `elif` expression that block uses `__thread`.

A simple way to test this is to add `-DF2PY_THREAD_LOCAL_DECL=__thread` to C/CXXFLAGS and see if the related warnings go away.

**Comment by carlkl** [comment]
Looking at the mingw-w64 sources it should be `__thread` for GCC and `__declspec(thread)` only for VS.

**Comment by pearu** [comment]
Thanks, @carlkl for looking into this. It sounds like `defined(__MINGW64__)` needs to be moved as suggested.
The corresponding code was introduced relatively recently: https://github.com/numpy/numpy/pull/18910

**Comment by rgommers** [comment]
Thanks @pearu and @carlkl! Submitted a PR.

**Comment by rgommers** [comment]
this was fixed in numpy 1.22.2, closing
